tools/runfiles/runfiles.rs

```diff
@@ -44,36 +44,49 @@ const MANIFEST_FILE_ENV_VAR: &str = "RUNFILES_MANIFEST_FILE";
 const MANIFEST_ONLY_ENV_VAR: &str = "RUNFILES_MANIFEST_ONLY";
 const TEST_SRCDIR_ENV_VAR: &str = "TEST_SRCDIR";
 
+#[macro_export]
+macro_rules! Rlocation {
+    ($r:ident, $path:expr) => {
+        $r.rlocation_from($path, env!("REPOSITORY_NAME"))
+    };
+}
+
 #[derive(Debug)]
 enum Mode {
     DirectoryBased(PathBuf),
     ManifestBased(HashMap<PathBuf, PathBuf>),
 }
 
+type RepoMappingKey = (String, String);
+type RepoMapping = HashMap<RepoMappingKey, String>;
+
 #[derive(Debug)]
 pub struct Runfiles {
     mode: Mode,
+    repo_mapping: RepoMapping,
 }
 
 impl Runfiles {
     /// Creates a manifest based Runfiles object when
     /// RUNFILES_MANIFEST_ONLY environment variable is present,
     /// or a directory based Runfiles object otherwise.
     pub fn create() -> io::Result<Self> {
-        if is_manifest_only() {
+        let mode = if is_manifest_only() {
             Self::create_manifest_based()
```

Comment:
IIRC @fmeum said that the canonical implementation is at [rules_python](https://github.com/bazelbuild/bazel/blob/master/tools/python/runfiles/runfiles.py) and that other implementations should seek to follow that. In all other implementations, we create the manifest based one if RUNFILES_MANIFEST_FILE exists, otherwise we create it based on RUNFILES_DIR.

I believe we should instead have:
```
let mode = if let Ok(manifest_file) = std::env::var(MANIFEST_FILE_ENV_VAR) {
  create_manifest_based(Path::from(manifest_file))?
} else {
  Mode::DirectoryBased(find_runfiles_dir()?)
}
```

Reply:
I'm happy to swap to that. It's a bit riskier since it's a behavior change but it's more correct you point out

Reply:
`RUNFILES_MANIFEST_ONLY` is a trap that breaks cross-platform builds and should not be used. Instead, I would say the following flow is the best you can do at this point in time:
1. If at least one of `RUNFILES_MANIFEST_FILE` or `RUNFILES_DIR` is set:
    1. If `RUNFILES_MANIFEST_FILE` is set and exists, use it.
    2. If `RUNFILES_DIR` is set and exists, use it.
    3. Error.
2. If the runfiles manifest exists in one of the well-known locations, use it.
3. If the runfiles directory exists in one of the well-known locations, use it.
4. Error.